### ISSUE TITLE
My reproduction results do not align with the results reported in the paper.

### ISSUE THREAD

**Issue by streamer-os** [s-sahoo/duo#9]
Thank you very much for sharing the code and checkpoints about the excellent work. After downloading the released checkpoints for AR, SEDD-A, Duo, and MDLM trained on OWT, I attempted to reproduce the zero-shot perplexity (PPL) results in Table 2 of the paper using the  models.

For example, I used the following command with the Duo checkpoint path :
python main.py \
  mode=ppl_eval \
  loader.batch_size=8 \
  loader.eval_batch_size=8 \
  data=openwebtext-split \
  model=small \
  algo=duo_base \
  eval.checkpoint_path=/share/kuleshov/ssahoo/flow-ode/flow-ode-VlCQLK-small-conjugate-OWT-anneal/checkpoints/last.ckpt \
  sampling.num_sample_batches=0 \
  +wandb.offline=true

I only changed the data field to evaluate on different datasets: PTB, Wikitext, Lambada, Pubmed, LM1B, AG News, and Arxiv. However, I found that the perplexity numbers I obtained are quite different from those reported in Table 2 of the paper.

Based on my understanding, is this the correct method to reproduce the results? Or are there additional steps or details I may be missing?

Also, when evaluating on the PTB dataset, I encountered an issue where a block_size of 1024 leads to an empty dataset, making evaluation impossible. How was this handled in your experiments?

Thanks in advance!

<img width="905" height="244" alt="Image" src="https://github.com/user-attachments/assets/29e5df55-8681-41ab-8be3-6ea9076be9f3" />

**Comment by s-sahoo** [comment]
You're missing `data.insert_valid_eos=False` flag. Please use the commands as specified in `zero_shot_{algo}.sh` scripts [here](https://github.com/s-sahoo/duo/tree/main/scripts). 

Furthermore, run these experiments on a single GPU, as I remember some small datasets had issues with multi-gpu eval, I forget which ones. 

Also, please use `duo.ckpt`

**Comment by streamer-os** [comment]
Thank you very much for your guidance. I’ve been able to mostly reproduce the results from the paper. However, when evaluating on the PTB dataset, I encountered an error using the provided command.

python main.py \
    mode=ppl_eval \
    loader.batch_size=16 \
    loader.eval_batch_size=16 \
    loader.eval_global_batch_size=128 \
    data=ptb \
    data.insert_valid_eos=False \
    model=small \
    algo=sedd\
    model.length=1024 \
    eval.checkpoint_path=sedd.ckpt\
    sampling.num_sample_batches=0 \
    +wandb.offline=true

After some debugging, I found that the issue is caused by the block_size parameter in group_texts. Specifically, setting block_size=1024 results in an empty dataset for PTB.

May I ask how you handled this issue during your evaluation? Did you use a different block size for PTB, or apply a different preprocessing approach?

<img width="542" height="146" alt="Image" src="https://github.com/user-attachments/assets/6e78e8d9-8612-4a2c-a6f6-8378293adb53" />

**Comment by s-sahoo** [comment]
> Re ptb


I was able to reproduce the zero-shot number for DUO on the `ptb` dataset as reported in the paper by using the command below. Probably your `ptb` dataset is corrupted. Could you specify a new path for the dataset to be downloaded and tokenized via `data.cache_dir`?

```
python -u -m main   \
mode=ppl_eval    \
loader.batch_size=16     \
loader.eval_batch_size=16     \
loader.eval_global_batch_size=128     \
data=ptb     \
data.insert_valid_eos=False     \
model=small     \
algo=duo_base     \
model.length=1024     \
eval.checkpoint_path=$checkpoint_path     \
sampling.num_sample_batches=0    \
 +wandb.offline=true \
data.cache_dir=/share/kuleshov/ssahoo/.trash/
```

**Comment by streamer-os** [comment]
> For AR, we took the numbers from [MDLM](https://arxiv.org/pdf/2406.07524). In my experience, a PPL difference of less than 0.3 is negligible. I’m not sure what’s going on with WikiText, though. Can you confirm that you’re running the evaluation on a single GPU? Also, could you compute the PPL evaluation with multiple seeds?
> 
> One of the MDLM co-authors who ran the zero-shot PPL numbers has graduated, but I’ll see if I can figure it out.

Thanks, for the AR evaluation, I did run it on a single GPU and also tried using a different random seed, but the results were exactly the same and still did not match the values reported in the paper.

**Comment by streamer-os** [comment]
> > Re ptb
> 
> I was able to reproduce the zero-shot number for DUO on the `ptb` dataset as reported in the paper by using the command below. Probably your `ptb` dataset is corrupted. Could you specify a new path for the dataset to be downloaded and tokenized via `data.cache_dir`?
> 
> ```
> python -u -m main   \
> mode=ppl_eval    \
> loader.batch_size=16     \
> loader.eval_batch_size=16     \
> loader.eval_global_batch_size=128     \
> data=ptb     \
> data.insert_valid_eos=False     \
> model=small     \
> algo=duo_base     \
> model.length=1024     \
> eval.checkpoint_path=$checkpoint_path     \
> sampling.num_sample_batches=0    \
>  +wandb.offline=true \
> data.cache_dir=/share/kuleshov/ssahoo/.trash/
> ```

I followed your instructions, but I'm still seeing the same error. After some investigation, I believe the issue lies in the _group_texts function in dataloader.py. The problem seems to be caused by the block_size being set to 1024, while the PTB dataset contains relatively short sentences, none of which reach that length. As a result, all examples are being filtered out. I’m wondering if this part of the code was modified on your side? Here is the function I’m referring to:

def _group_texts(examples, block_size, bos, eos): 
  # Concatenate all texts.
  concatenated_examples = list(itertools.chain(* examples['input_ids']))
  total_length = len(concatenated_examples)

  # TODO(yair): look into not dropping the remainder but rather padding it.
  # We drop the small remainder, and if the total_length < block_size - 2
  # we exclude this batch and return an empty dict.
  # We could add padding if the model supported it instead of
  # this drop, you can customize this part to your needs.
  new_block_size = block_size - 2  # [BOS] and [EOS] to be added
  total_length = (total_length // new_block_size) * new_block_size

  # Split by chunks of max_len.
  result = {}
  _values = []
  _attn_masks = []
  for i in range(0, total_length, new_block_size):
    _values.append(
      [bos]
      + concatenated_examples[i : i + new_block_size]
      + [eos])
    _attn_masks.append(torch.ones(block_size))
  result['input_ids'] = _values
  result['attention_mask'] = _attn_masks
  return result

**Comment by s-sahoo** [comment]
There’s nothing wrong with the codebase—I’m using the same code released in this repository. It seems to me that you’re not downloading the complete PTB dataset, which is likely causing this error. Unfortunately, I’m not able to offer much help beyond this.

Regarding AR: At this point, it’s best if you recompute the new numbers for all the methods in your paper.

**Comment by streamer-os** [comment]
> There’s nothing wrong with the codebase—I’m using the same code released in this repository. It seems to me that you’re not downloading the complete PTB dataset, which is likely causing this error. Unfortunately, I’m not able to offer much help beyond this.
> 
> Regarding AR: At this point, it’s best if you recompute the new numbers for all the methods in your paper.

Thank you very much! I’ll give it another try.
